simulator/infra/ports/t/topology_test.cpp

```diff
@@ -0,0 +1,146 @@
+            from_D = make_read_port<int>( "D_to_A", PORT_LATENCY);
+        }
+    };
+    struct B : public Module {
+        std::unique_ptr<WritePort<int>> to_C;
+        std::unique_ptr<WritePort<int>> to_D;
+        std::unique_ptr<ReadPort<int>> from_C;
+        std::unique_ptr<ReadPort<int>> from_D;
+        explicit B( Module* root) : Module( root, "B")
+        {
+            to_C = make_write_port<int>( "B_to_C", PORT_BW);
+            to_D = make_write_port<int>( "B_to_D", PORT_BW);
+            from_C = make_read_port<int>( "C_to_B", PORT_LATENCY);
+            from_D = make_read_port<int>( "D_to_B", PORT_LATENCY);
+        }
+    };
+    struct C : public Module {
+        std::unique_ptr<WritePort<int>> to_A;
+        std::unique_ptr<WritePort<int>> to_B;
+        std::unique_ptr<ReadPort<int>> from_A;
+        std::unique_ptr<ReadPort<int>> from_B;
+        explicit C( Module* root) : Module( root, "C")
+        {
+            to_A = make_write_port<int>( "C_to_A", PORT_BW);
+            to_B = make_write_port<int>( "C_to_B", PORT_BW);
+            from_A = make_read_port<int>( "A_to_C", PORT_LATENCY);
+            from_B = make_read_port<int>( "B_to_C", PORT_LATENCY);
+        }
+    };
+    struct D : public Module {
+        std::unique_ptr<WritePort<int>> to_A;
+        std::unique_ptr<WritePort<int>> to_B;
+        std::unique_ptr<ReadPort<int>> from_A;
+        std::unique_ptr<ReadPort<int>> from_B;
+        explicit D( Module* root) : Module( root, "D")
+        {
+            to_A = make_write_port<int>( "D_to_A", PORT_BW);
+            to_B = make_write_port<int>( "D_to_B", PORT_BW);
+            from_A = make_read_port<int>( "A_to_D", PORT_LATENCY);
+            from_B = make_read_port<int>( "B_to_D", PORT_LATENCY);
+        }
+    };
+    std::unique_ptr<A> a;
+    std::unique_ptr<B> b;
+    std::unique_ptr<C> c;
+    std::unique_ptr<D> d;
+    void modulemap_load( pt::ptree* modulemap) const { modulemap_dumping( modulemap); }
+    void portmap_load( pt::ptree* portmap) const { portmap_dumping( portmap); }
+    void modules_load( pt::ptree* modules) const { module_dumping( modules); }
+    void topology_load( pt::ptree* topology) const { topology_dumping_impl( topology); }
+    void topology_save() { topology_dumping(true); }
+    void topology_save(const std::string& filename) { topology_dumping(true, filename); }
+    bool check_if_dumps() const { return sout.enabled(); }
+    explicit SomeTopology()
+    {
+        /* Root
+         * | \
+         * A  C
+         * |   \ 
+         * B    D  
+         */
+        a = std::make_unique<A>( this);
+        c = std::make_unique<C>( this);
+        b = std::make_unique<B>( a.get());
+        d = std::make_unique<D>( c.get());
+        init_portmap();
+    }
+};
+
+TEST_CASE( "Topology: modulemap")
+{
+    SomeTopology t;
+    pt::ptree exp_modulemap;
+    pt::ptree modulemap;
+    read_json( TEST_PATH + std::string( "/modulemap_test.json"), exp_modulemap);

```

Comment:
Can we extract `pt::tree` ctor and `read_json` to a new method, i.e.

```c++
static auto read_json_from_file(const std::string& path)
{
     // ...
     return result;
}
```

That would reduce level of copy-paste.